internal/analyzers/overlapping.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2024 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analyzers
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"strings"
+
+	"github.com/prometheus-operator/poctl/internal/k8sutil"
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func RunOverlappingAnalyzer(ctx context.Context, clientSets *k8sutil.ClientSets, _, namespace string) error {
+	serviceMonitors, err := clientSets.MClient.MonitoringV1().ServiceMonitors(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	podMonitors, err := clientSets.MClient.MonitoringV1().PodMonitors(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	if (serviceMonitors == nil || len(serviceMonitors.Items) == 0) && (podMonitors == nil || len(podMonitors.Items) == 0) {
+		return nil
+	}
+
+	serviceOverlaps := make(map[string][]string)
+	podOverlaps := make(map[string][]string)
+	var overlapErrs []string
+
+	for _, servicemonitor := range serviceMonitors.Items {
+		if err := checkOverlappingServiceMonitors(ctx, clientSets, servicemonitor, serviceOverlaps); err != nil {
+			overlapErrs = append(overlapErrs, err.Error())
+		}
+	}
+	for _, podmonitor := range podMonitors.Items {
+		if err := checkOverlappingPodMonitors(ctx, clientSets, podmonitor, podOverlaps); err != nil {
+			overlapErrs = append(overlapErrs, err.Error())
+		}
+	}
+
+	for key, svcMonitors := range serviceOverlaps {
+		if len(svcMonitors) > 1 {
+			overlapErrs = append(overlapErrs, fmt.Sprintf("Overlapping ServiceMonitors found for service/port %s: %v", key, svcMonitors))
+		}
+	}
+
+	for key, pdMonitors := range podOverlaps {
```

Comment:
If we properly build the error message inside each for we don't nee to iterate it again to rebuild the error mesage, wdyt?

Reply:
Hi, mv the error inside the overlapping function, wdyt?